### PR TITLE
raftstore: clean stale peer based on oldest snapshot sequence

### PR DESCRIPTION
**Pull request by Connor1996**
###  What have you changed?
Now TiKV delays `delete_files_in_range_cf` about 10min(it also is the max txn time for tidb) when destroying peer to keep the consistency of existing snapshots used by readpool.

Once the oldest snapshot sequence exceeds that sequence when destroying peer, it means no read request is reading the data of the destroyed peer. So we can safely call `delete_files_in_range_cf` at that time. 

After this PR, we can accelerate the speed of space reclamation dramatically.

###  What is the type of the changes?
- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?
- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?
No

###  Does this PR affect `tidb-ansible`?
No

###  Refer to a related PR or issue link (optional)
https://github.com/tikv/tikv/issues/6200

### Release note 
- clean stale peer based on oldest snapshot sequence

### CODE REVIEW

**Comment by siddontang** [comment]
> After this PR, we can accelerate the speed of space reclamation dramatically.

This is cool, can you show the result to represent it?

**Comment by Connor1996** [comment]
@lhy1024 will use this PR to test the effect on PD balance algorithm, please paste the result here by the way. Thanks~

**Comment by zhangjinpeng1987** [comment]
Please fix conflicts.

File: Cargo.lock [hunk 1]
```diff
@@ -1471,7 +1471,7 @@ dependencies = [
 [[package]]
 name = "librocksdb_sys"
 version = "0.1.0"
-source = "git+https://github.com/pingcap/rust-rocksdb.git#4ad6d0a7b35d30e7c67f6cfdff7d97874f330123"
+source = "git+https://github.com/Connor1996/rust-rocksdb.git?branch=snapshot-oldest#a0e6e48636b9a4a21cba4737534f8d3fc49984a6"
```
Comment (zhangjinpeng1987) [inline Cargo.lock]: Don't forget to update this dependency.

**Comment by siddontang** [comment]
image

@Connor1996 @lhy1024 

How about the result?

**Comment by Connor1996** [comment]
before: store used goes down with a delay of 10min compring to region size
屏幕快照 2019-12-25 下午2 22 21

after: store used goes down almost synchronously with region size's decline
屏幕快照 2019-12-25 下午2 22 30

**Comment by Connor1996** [comment]
@NingLin-P  Thanks a lot!

File: components/raftstore/src/store/worker/region.rs [hunk 2]
```diff
@@ -449,23 +455,29 @@ where
...
-        let use_delete_files = false;
-        for (region_id, s_key, e_key) in overlap_ranges {
+        if overlap_ranges.is_empty() {
+            return;
+        }
+        let oldest_sequence = self
+            .engines
+            .kv
+            .get_oldest_snapshot_sequence_number()
+            .unwrap_or(u64::MAX);
```
Comment (Connor1996) [inline components/raftstore/src/store/worker/region.rs:465]: if no snapshot is in use, I think we can safely call `delete_files_in_range`

File: components/raftstore/src/store/worker/region.rs [hunk 3]
```diff
@@ -449,23 +455,29 @@ where
...
+        for (region_id, s_key, e_key, stale_sequence) in overlap_ranges {
+            // `delete_files_in_range` may break current rocksdb snapshots consistency,
+            // so do not use it unless we can make sure there is no reader of the destroyed peer anymore.
+            let use_delete_files = stale_sequence < oldest_sequence;
+            if use_delete_files {
```
Comment (Connor1996) [inline components/raftstore/src/store/worker/region.rs:470]:
```suggestion
            if !use_delete_files {
```

File: components/raftstore/src/store/worker/region.rs [hunk 4]
```diff
@@ -449,23 +455,29 @@ where
...
+                SNAP_COUNTER_VEC
+                    .with_label_values(&["overlap", "not_timeout"])
```
Comment (Connor1996) [inline components/raftstore/src/store/worker/region.rs:472]:
```suggestion
                    .with_label_values(&["apply", "not_delete_files"])
```

File: components/raftstore/src/store/worker/region.rs [hunk 5]
```diff
@@ -449,23 +455,29 @@ where
...
+        }
+        let oldest_sequence = self
+            .engines
+            .kv
+            .get_oldest_snapshot_sequence_number()
+            .unwrap_or(u64::MAX);
+        for (region_id, s_key, e_key, stale_sequence) in overlap_ranges {
+            // `delete_files_in_range` may break current rocksdb snapshots consistency,
+            // so do not use it unless we can make sure there is no reader of the destroyed peer anymore.
+            let use_delete_files = stale_sequence < oldest_sequence;
```
Comment (Connor1996) [inline components/raftstore/src/store/worker/region.rs:469]: seems I was wrong. Good job @NingLin-P

File: components/raftstore/src/store/worker/region.rs (hunk 2 above)
Comment (NingLin-P) [reply components/raftstore/src/store/worker/region.rs:465]: Yes, by setting `oldest_sequence` to `u64::MAX`,  `use_delete_files` always be `true` and  `delete_files_in_range ` will be called.

**Comment by Connor1996** [comment]
PTAL @overvenus @zhangjinpeng1987

**Pull request merged**